### PR TITLE
Fix subtracted background bug in outlier_detection

### LINKED ISSUE #4494: Outlier Detection step for calwebb_coron3 crashes
Issue [JP-1251](https://jira.stsci.edu/browse/JP-1251) was created by Brendan Hagan:

I am testing calwebb_coron3 with the data found here: [https://stsci.box.com/s/79isxou69rzltpal36fm2uxizofljqw3]

The ASN file ('test.yml') points to the three FITS files in the box folder - I specify one target (i.e. exptype: science) and two references (i.e. exptype: psf). Each FITS file contains 5 integrations (i.e. 5 slope images in each) for the MIRI 1065 4QPM. Each FITS file went through Detector1Pipeline and Image2Pipeline steps.

To generate the bug, I run the following lines in a jupyter notebook:
{code:java}
from jwst.pipeline import Coron3Pipeline
data_dir = '[directory location of box folder containing data]'
asn_dir = data_dir+'test.yml'
myCoron3Pipeline = Coron3Pipeline()
myCoron3Pipeline.save_results = True
myCoron3Pipeline.align_refs.override_psfmask = data_dir + 'psfmask_MIRI_4QPM_1140.fits'
myCoron3Pipeline.output_dir = data_dir
myCoron3Pipeline.run(asn_dir)
{code}
The full output is here:
{code:python}
2020-01-29 10:13:39,212 - stpipe.Coron3Pipeline - INFO - Coron3Pipeline instance created.
2020-01-29 10:13:39,213 - stpipe.Coron3Pipeline.stack_refs - INFO - StackRefsStep instance created.
2020-01-29 10:13:39,215 - stpipe.Coron3Pipeline.align_refs - INFO - AlignRefsStep instance created.
2020-01-29 10:13:39,216 - stpipe.Coron3Pipeline.klip - INFO - KlipStep instance created.
2020-01-29 10:13:39,219 - stpipe.Coron3Pipeline.outlier_detection - INFO - OutlierDetectionStep instance created.
2020-01-29 10:13:39,221 - stpipe.Coron3Pipeline.resample - INFO - ResampleStep instance created.
2020-01-29 10:13:39,269 - stpipe.Coron3Pipeline - INFO - Step Coron3Pipeline running with args ('/Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub/test.yml',).
2020-01-29 10:13:39,275 - stpipe.Coron3Pipeline - INFO - Step Coron3Pipeline parameters are: {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': '/Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub/', 'output_ext': '.fits', 'output_use_model': False, 'output_use_index': True, 'save_results': True, 'skip': False, 'suffix': 'i2d', 'search_output_file': True, 'input_dir': '', 'stack_refs': {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.fits', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': True, 'input_dir': ''}, 'align_refs': {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.fits', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': True, 'input_dir': ''}, 'klip': {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.fits', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': True, 'input_dir': '', 'truncate': 50}, 'outlier_detection': {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.fits', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': False, 'input_dir': '', 'weight_type': 'exptime', 'pixfrac': 1.0, 'kernel': 'square', 'fillval': 'INDEF', 'nlow': 0, 'nhigh': 0, 'maskpt': 0.7, 'grow': 1, 'snr': '4.0 3.0', 'scale': '0.5 0.4', 'backg': 0.0, 'save_intermediate_results': False, 'resample_data': True, 'good_bits': 6, 'scale_detection': False}, 'resample': {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.fits', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': True, 'input_dir': '', 'pixfrac': None, 'kernel': None, 'fillval': None, 'weight_type': None, 'good_bits': 6, 'single': False, 'blendheaders': True}}
2020-01-29 10:13:39,276 - stpipe.Coron3Pipeline - INFO - Starting calwebb_coron3 ...
202001291013:DEBUG:jwst.associations.association_io.load:Error unserializing: "Expecting property name enclosed in double quotes: line 12 column 11 (char 363)"
2020-01-29 10:13:39,856 - stpipe.Coron3Pipeline - INFO - Prefetching reference files for dataset: 'MIRI_5715_141_S_20191106-000341_SCE1_pipe_edited_calints.fits' reftypes = ['drizpars']
2020-01-29 10:13:41,261 - stpipe.Coron3Pipeline - INFO - Prefetch for DRIZPARS reference file is '/grp/crds/cache/references/jwst/jwst_miri_drizpars_0001.fits'.
2020-01-29 10:13:41,269 - stpipe.Coron3Pipeline - INFO - Override for PSFMASK reference file is '/Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub/psfmask_MIRI_4QPM_1140.fits'.
2020-01-29 10:13:41,499 - stpipe.Coron3Pipeline - INFO - Prefetching reference files for dataset: 'MIRI_5717_141_S_20191106-224812_SCE1_pipe_edited_calints.fits' reftypes = ['drizpars']
2020-01-29 10:13:41,501 - stpipe.Coron3Pipeline - INFO - Prefetch for DRIZPARS reference file is '/grp/crds/cache/references/jwst/jwst_miri_drizpars_0001.fits'.
2020-01-29 10:13:41,512 - stpipe.Coron3Pipeline - INFO - Override for PSFMASK reference file is '/Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub/psfmask_MIRI_4QPM_1140.fits'.
2020-01-29 10:13:41,740 - stpipe.Coron3Pipeline - INFO - Prefetching reference files for dataset: 'MIRI_5715_61_S_20191105-202437_SCE1_pipe_edited_calints.fits' reftypes = ['drizpars']
2020-01-29 10:13:41,743 - stpipe.Coron3Pipeline - INFO - Prefetch for DRIZPARS reference file is '/grp/crds/cache/references/jwst/jwst_miri_drizpars_0001.fits'.
2020-01-29 10:13:41,753 - stpipe.Coron3Pipeline - INFO - Override for PSFMASK reference file is '/Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub/psfmask_MIRI_4QPM_1140.fits'.
2020-01-29 10:13:42,250 - stpipe.Coron3Pipeline.stack_refs - INFO - Step stack_refs running with args (<ModelContainer>,).
2020-01-29 10:13:42,252 - stpipe.Coron3Pipeline.stack_refs - INFO - Step stack_refs parameters are: {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.fits', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': True, 'input_dir': '/Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub'}
2020-01-29 10:13:42,259 - stpipe.Coron3Pipeline.stack_refs - INFO -  Adding psf member 1 to output stack
2020-01-29 10:13:42,276 - stpipe.Coron3Pipeline.stack_refs - INFO -  Adding psf member 2 to output stack
2020-01-29 10:13:42,325 - stpipe.Coron3Pipeline.stack_refs - INFO - Step stack_refs done
2020-01-29 10:13:42,453 - stpipe.Coron3Pipeline - INFO - Saved model in /Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub/jw10005-miri-mask1065_psfstack.fits
2020-01-29 10:13:42,512 - stpipe.Coron3Pipeline.align_refs - INFO - Step align_refs running with args ('/Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub/MIRI_5715_61_S_20191105-202437_SCE1_pipe_edited_calints.fits', <CubeModel(10, 224, 288) from jw10005-miri-mask1065_psfstack.fits>).
2020-01-29 10:13:42,513 - stpipe.Coron3Pipeline.align_refs - INFO - Step align_refs parameters are: {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.fits', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': True, 'input_dir': '/Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub'}
2020-01-29 10:13:42,805 - stpipe.Coron3Pipeline.align_refs - INFO - Using PSFMASK reference file /Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub/psfmask_MIRI_4QPM_1140.fits
2020-01-29 10:14:00,049 - stpipe.Coron3Pipeline.align_refs - INFO - Step align_refs done
2020-01-29 10:14:00,303 - stpipe.Coron3Pipeline - INFO - Saved model in /Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub/MIRI_5715_61_S_20191105-202437_SCE1_pipe_edited_c1001_psfalign.fits
2020-01-29 10:14:00,361 - stpipe.Coron3Pipeline.klip - INFO - Step klip running with args ('/Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub/MIRI_5715_61_S_20191105-202437_SCE1_pipe_edited_calints.fits', <QuadModel(5, 10, 224, 288) from MIRI_5715_61_S_20191105-202437_SCE1_pipe_edited_c1001_psfalign.fits>).
2020-01-29 10:14:00,363 - stpipe.Coron3Pipeline.klip - INFO - Step klip parameters are: {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.fits', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': True, 'input_dir': '/Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub', 'truncate': 50}
2020-01-29 10:14:00,615 - stpipe.Coron3Pipeline.klip - INFO - KL transform truncation = 50
2020-01-29 10:14:00,931 - stpipe.Coron3Pipeline.klip - INFO - Step klip done
2020-01-29 10:14:01,142 - stpipe.Coron3Pipeline - INFO - Saved model in /Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub/MIRI_5715_61_S_20191105-202437_SCE1_pipe_edited_c1001_psfsub.fits
2020-01-29 10:14:01,363 - stpipe.Coron3Pipeline.outlier_detection - INFO - Step outlier_detection running with args (<ModelContainer>,).
2020-01-29 10:14:01,366 - stpipe.Coron3Pipeline.outlier_detection - INFO - Step outlier_detection parameters are: {'pre_hooks': [], 'post_hooks': [], 'output_file': None, 'output_dir': None, 'output_ext': '.fits', 'output_use_model': False, 'output_use_index': True, 'save_results': False, 'skip': False, 'suffix': None, 'search_output_file': False, 'input_dir': '/Users/hagan/Master_Folder/James_Hagan/STScI/Functional/panCAKE/pancake_pipeline_simulations/jpl9_sub', 'weight_type': 'exptime', 'pixfrac': 1.0, 'kernel': 'square', 'fillval': 'INDEF', 'nlow': 0, 'nhigh': 0, 'maskpt': 0.7, 'grow': 1, 'snr': '4.0 3.0', 'scale': '0.5 0.4', 'backg': 0.0, 'save_intermediate_results': False, 'resample_data': True, 'good_bits': 6, 'scale_detection': False}
2020-01-29 10:14:01,370 - stpipe.Coron3Pipeline.outlier_detection - INFO - Performing outlier detection on 5 inputs
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-4-3a70ab7d1ec7> in <module>
      4 myCoron3Pipeline.align_refs.override_psfmask = data_dir + 'psfmask_MIRI_4QPM_1140.fits'
      5 myCoron3Pipeline.output_dir = data_dir
----> 6 myCoron3Pipeline.run(asn_dir)
      7 # myCoron3Pipeline.outlier_detection.skip = True
      8 # myCoron3Pipeline.resample.skip = True

~/anaconda3/lib/python3.7/site-packages/jwst/stpipe/step.py in run(self, *args)
    389                     self.prefetch(*args)
    390                 try:
--> 391                     step_result = self.process(*args)
    392                 except TypeError as e:
    393                     if "process() takes exactly" in str(e):

~/anaconda3/lib/python3.7/site-packages/jwst/pipeline/calwebb_coron3.py in process(self, input)
    132 
    133             # Call outlier_detection
--> 134             target_models = self.outlier_detection(target_models)
    135 
    136             # Create Level 2c products

~/anaconda3/lib/python3.7/site-packages/jwst/stpipe/step.py in run(self, *args)
    389                     self.prefetch(*args)
    390                 try:
--> 391                     step_result = self.process(*args)
    392                 except TypeError as e:
    393                     if "process() takes exactly" in str(e):

~/anaconda3/lib/python3.7/site-packages/jwst/outlier_detection/outlier_detection_step.py in process(self, input)
    164             # Set up outlier detection, then do detection
    165             step = detection_step(self.input_models, reffiles=reffiles, **pars)
--> 166             step.do_detection()
    167 
    168             if self.input_container:

~/anaconda3/lib/python3.7/site-packages/jwst/outlier_detection/outlier_detection.py in do_detection(self)
    245         # Perform outlier detection using statistical comparisons between
    246         # each original input image and its blotted version of the median image
--> 247         self.detect_outliers(blot_models)
    248 
    249         # clean-up (just to be explicit about being finished with

~/anaconda3/lib/python3.7/site-packages/jwst/outlier_detection/outlier_detection.py in detect_outliers(self, blot_models)
    347 
    348         for image, blot in zip(self.input_models, blot_models):
--> 349             flag_cr(image, blot, **self.outlierpars)
    350 
    351         if self.converted:

~/anaconda3/lib/python3.7/site-packages/jwst/outlier_detection/outlier_detection.py in flag_cr(sci_image, blot_image, **pars)
    391         subtracted_background = sci_image.meta.background.level
    392         log.debug("Subtracted background: {}".format(subtracted_background))
--> 393     if subtracted_background is None:
    394         subtracted_background = backg
    395 

UnboundLocalError: local variable 'subtracted_background' referenced before assignment
{code}
Indeed, looking at line 393 in outlier_detection.py, it seems that 'subtracted_background' variable is never defined before being referenced.

### PR DESCRIPTION
**Pull request by jdavies-st**
Resolves #4494 and [JP-1251](https://jira.stsci.edu/browse/JP-1251)

### CODE REVIEW

File: jwst/outlier_detection/outlier_detection.py [hunk 1]
```diff
@@ -386,11 +386,14 @@ def flag_cr(sci_image, blot_image, **pars):
     snr1, snr2 = [float(val) for val in pars.get('snr', '5.0 4.0').split()]
     scl1, scl2 = [float(val) for val in pars.get('scale', '1.2 0.7').split()]
 
-    if not sci_image.meta.background.subtracted:
-        # Include background back into blotted image for comparison
+    # Include subtracted background back into blotted image for comparison
+    if (sci_image.meta.background.subtracted is False and
```
Comment (hbushouse) [inline jwst/outlier_detection/outlier_detection.py]: I'm confused. If ``background.subtracted`` is False that would imply that no background has been subtracted from this image (i.e. the background signal is still in the image), so there shouldn't be any need to add it back in (it's already there).

Comment (jdavies-st) [reply jwst/outlier_detection/outlier_detection.py]: Agree.  I suspect the logic is incorrect.  Fixing.

Maybe @stsci-hack has thoughts on this?

Are we squashing two bugs for the price of one here?

Comment (stsci-hack) [reply jwst/outlier_detection/outlier_detection.py]: @hbushouse is right, and the changes look like they are more logical than what was there before.  Unfortunately, I can't remember why it was implemented that way in the first place...

**Comment by jdavies-st** [comment]
Added a test because nothing in `outlier_detection` is unit tested, and `flag_cr` is the easiest (only easy?) part to test.  The test doesn't actually test if the noise model is correct for JWST data, only that the updated logic in this PR works, and that the DQ array gets updated with a non-zero value when there's an egregious outlier.

The flag set by this step currently is `4, "JUMP_DET"`.  That is incorrect. It should be `1, "DO_NOT_USE"`.  When this is fixed in a future PR (and the downstream changes in resample are also done at the same time), then these unit tests can be updated to check the correct bit is set in the DQ array.

There should also be tests written that actually make sure the noise model actually works for JWST multi-read data as well.  Not for this PR.

**Pull request merged**